tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -238,16 +291,61 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     return Status::OK();
   }
 
+  Status HandleMaximum(HloInstruction *instr) override {
+    HloInstruction *existing_gemm, *slice;
+    auto zero_pattern = m::Broadcast(m::ConstantScalar(0));
+    // Attempt to elide maximum and fuse ReLU activation into GEMM.
+    if (Match(instr, m::MaximumAnyOrder(zero_pattern,
+                                        m::Op(&existing_gemm)
+                                            .WithCustomCallTarget(
+                                                kCublasLtMatmulCallTarget)))) {
+      TF_ASSIGN_OR_RETURN(bool was_fused,
+                          FuseReluActivationSliced(instr, existing_gemm));
+      if (was_fused) {
+        return OkStatus();
+      }
+    }
+    // Attempt to elide maximum and bitcast and fuse ReLU activation into
+    // *batched* GEMM.
+    if (Match(instr, m::MaximumAnyOrder(
+                         zero_pattern,
+                         m::Bitcast(m::Op(&existing_gemm)
```

Comment:
Since we probably don't want to swap for any possible `f`, it seems that we would have to extend the functionality to match a prescribed set of ops (e.g. Add and Max) and their respective operands before we could reduce the number of patterns.

I've introduced short forms of the ops that often appear in the patterns which makes them a bit less complex and error-prone.

Reply:
The change doesn't make the implementation measurably shorter, but it might reduce the complexity and potential for errors by separating the definition of the ops from the matching. What I meant above is what you write as well, i.e. that  further streamlining of the matching would likely require an extension of the functionality of the pattern matcher.

Reply:
> What I meant above is what you write as well, i.e. that further streamlining of the matching would likely require an extension of the functionality of the pattern matcher.

Yeah.  As we add more complex pattern-matches, it's not super-surprising to me that we'd need to extend the pattern-matcher a bit.

Reply:
In my view, an extension of the functionality of the pattern matcher exceeds the scope of this PR (which was already approved), but could be introduced separately. The short forms of the ops are now class members which avoids their duplication.